### DISCUSSION TITLE
batch processing

### DISCUSSION THREAD

**Discussion by gour** [ad-si/TaskLite#83] [Q&A]
Hello!

Let me first congratulate you for 0.4 release!

It brings, afaict, proper support for recurring tasks which is feature not addressed even in Taskwarrior-3.x and which could even persuade me to switch to TL!. When it comes to contributing, I also know that I'd not learn Rust or go back to C++, but Haskell is the most beautiful language I've ever encountered...tried to learn it twice, but didn't grok monads. :-)

Several times I was trying to replace TW with Emacs's org-mode, but always returning to TW or, recently,. using [Etm](https://github.com/dagraham/etm-dgraham) which is very nice and complete, but lacks the elegance of TW's cli interface.

I did import TW's tasks, but I wonder how to be more effective with TL which uses more complex ids for tasks in comparison with TW.

For instance, after fresh import of TW tasks, there are many which are obsolete or overdue, so I wonder how could I 'do' them all, iow. something like:

`tasklite overdue | 'do them all'

Sincerely,
Gour

**Comment by ad-si** [comment]
Thanks! 😊
Agreed, Haskell is awesome! Don't get scared by monads. You don't really need to understand them to use them and write usable Haskell code. Just think of them as "context": If you're in the `IO` monad, you're in the `IO` context - aka you can write/read files & databases. That's all there is to know to get started.

To script the TaskLite CLI the easiest way is often to run a raw SQL command:

```bash
tasklite runsql 'select ulid from tasks where due_utc < datetime()' \
| tail -n +2 \
| while read -r taskId; do tasklite do "$taskId"; done
```

**Reply by gour** [reply]
> Agreed, Haskell is awesome! Don't get scared by monads. You don't really need to understand them to use them and write usable Haskell code. Just think of them as "context": If you're in the `IO` monad, you're in the `IO` context - aka you can write/read files & databases. That's all there is to know to get started.

I believe I can understand them conceptually as a way to isolate pure code from side-effects...wonder if you had a need to design/implement your own monad or the library provided is good-enough?

Which editor do you use/recommend for writing Haskell code?

> To script the TaskLite CLI the easiest way is often to run a raw SQL command:
> 
> ```shell
> tasklite runsql 'select ulid from tasks where due_utc < datetime()' \
> | tail -n +2 \
> | while read -r taskId; do tasklite do "$taskId"; done
> ```

I'm still experimenting, so I did run the above shell cmd althought not fully understanding it, but I got:

```
$ tasklite runsql 'select ulid from tasks where due_utc < datetime()' \
> | tail -n +2 \
> | while read -r taskId; do tasklite do "$taskId"; done
tasklite: sqlite3: readCreateProcess: posix_spawnp: does not exist (No such file or directory)
```

Otoh, afaict, missing parts (in comparison wth TW) could be implemented via shell aliases, right?

**Reply by ad-si** [reply]
> I believe I can understand them conceptually as a way to isolate pure code from side-effects...wonder if you had a need to design/implement your own monad or the library provided is good-enough?

No need to over-complicate it. The normal `IO` monad is enough for the time being.

> Which editor do you use/recommend for writing Haskell code?

VS Code with the [official Haskell extension](https://marketplace.visualstudio.com/items?itemName=haskell.haskell) is pretty solid nowadays.

> tasklite: sqlite3: readCreateProcess: posix_spawnp: does not exist (No such file or directory)

The issue is, that sqlite3 isn't installed globally and you need to install it first. Should be available with the package manager of your OS. Which OS are you using?

But actually this should better also work without a global installation. I created an issue for it: https://github.com/ad-si/TaskLite/issues/86

> Otoh, afaict, missing parts (in comparison wth TW) could be implemented via shell aliases, right?

Correct! I tried to make it easily scriptable / extendable, so that missing features can be easily implemented with a few lines of bash (or any other scripting language).

**Reply by gour** [reply]
> No need to over-complicate it. The normal `IO` monad is enough for the time being.

:-)
> VS Code with the [official Haskell extension](https://marketplace.visualstudio.com/items?itemName=haskell.haskell) is pretty solid nowadays.

Thanks!

> The issue is, that sqlite3 isn't installed globally and you need to install it first. Should be available with the package manager of your OS.

Ahh, didn't think about it - it works now.
 
> Which OS are you using?

openSUSE Tumbleweed.

> Correct! I tried to make it easily scriptable / extendable, so that missing features can be easily implemented with a few lines of bash (or any other scripting language).

I like that approach. ;-)

**Comment by gour** [comment]
I get same error when just running:

```$ tl sql
tasklite: sqlite3: readCreateProcess: posix_spawnp: does not exist (No such file or directory)
```

Any hint?

**Reply by ad-si** [reply]
The issue is, that sqlite3 isn't installed globally and you need to install it first. Should be available with the package manager of your OS. Which OS are you using?
